router/qrouter/proxy_routing.go

```diff
@@ -158,15 +159,23 @@ func (qr *ProxyQrouter) DeparseKeyWithRangesInternal(ctx context.Context, key st
 	return nil, ComplexQuery
 }
 
-func (qr *ProxyQrouter) RouteKeyWithRanges(ctx context.Context, expr lyx.Node, meta *RoutingMetadataContext) (*routingstate.DataShardRoute, error) {
+func (qr *ProxyQrouter) RouteKeyWithRanges(ctx context.Context, expr lyx.Node, meta *RoutingMetadataContext, hf hashfunction.HashFunctionType) (*routingstate.DataShardRoute, error) {
 	switch e := expr.(type) {
 	case *lyx.ParamRef:
 		if e.Number >= len(meta.params) {
 			return nil, ComplexQuery
 		}
-		return qr.DeparseKeyWithRangesInternal(ctx, string(meta.params[e.Number]), meta)
+		hashedKey, err := hashfunction.ApplyHashFunction((meta.params[e.Number]), hf)
+		if err != nil {
+			return nil, err
+		}
+		return qr.DeparseKeyWithRangesInternal(ctx, string(hashedKey), meta)
 	case *lyx.AExprConst:
-		return qr.DeparseKeyWithRangesInternal(ctx, e.Value, meta)
+		hashedKey, err := hashfunction.ApplyHashFunction([]byte(e.Value), hf)
+		if err != nil {
+			return nil, err
+		}
+		return qr.DeparseKeyWithRangesInternal(ctx, string(hashedKey), meta)
```

Comment:
You can move key hashing before the switch

Reply:
here we cannot - look closer, meta.params[e.Number] vs e.Value